File: ASSIGNMENT_PCA_Wine.py
############ ASSIGNMENT PCA using wine dataset

# Business Problem: To perform dimension reduction using principal component analysis

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.decomposition import PCA
from sklearn.preprocessing import scale

# Data set details
wine_data = pd.read_csv("E:/PCA/wine.csv")
wine_data.shape # 178 obs, 14 variables
wine_data.columns
''' 'Type', 'Alcohol', 'Malic', 'Ash', 'Alcalinity', 'Magnesium', 'Phenols',
       'Flavanoids', 'Nonflavanoids', 'Proanthocyanins', 'Color', 'Hue',
       'Dilution', 'Proline' '''
wine_data.info() # all are continuous data

from collections import Counter
Counter(wine_data.Type)
# Type is categorical data with 3 classes

wine_data.describe()

# let us remove variable 'Type' 
wine = wine_data.iloc[:,1:]
wine.head()

# normalize the data
wine_normal = scale(wine)

# Model building
pca = PCA()
pca_scores = pca.fit_transform(wine_normal)
pca_scores = pd.DataFrame(pca_scores)
pca_scores.columns=['PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','PC11','PC12','PC13']
pca_scores.shape 
pca_scores.head()
# 178, 13  same number of Principal components as the variables

pca_comp = pd.DataFrame([pca.components_[0],pca.components_[1],pca.components_[2],
                         pca.components_[3],pca.components_[4],pca.components_[5],
                         pca.components_[6],pca.components_[7],pca.components_[8],
                         pca.components_[9],pca.components_[10],pca.components_[11],
                         pca.components_[12]])

pca_var = pd.Series(['Alcohol', 'Malic', 'Ash', 'Alcalinity', 'Magnesium', 'Phenols',
       'Flavanoids', 'Nonflavanoids', 'Proanthocyanins', 'Color', 'Hue',
       'Dilution', 'Proline'])

# pca_comp = pca_comp.transpose()
pca_comp = pd.concat([pca_var,pca_comp],axis=1)
pca_comp.columns = ['Variables','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','PC11','PC12','PC13']
pca_comp

# to calculate pca scores (using standr variabes and PC) in excel
pca_comp.to_excel(r'E:\PCA\pca.xlsx')
wine_normal = pd.DataFrame(wine_normal)
wine_normal.columns=['Alcohol', 'Malic', 'Ash', 'Alcalinity', 'Magnesium', 'Phenols',
       'Flavanoids', 'Nonflavanoids', 'Proanthocyanins', 'Color', 'Hue',
       'Dilution', 'Proline']
wine_normal = wine_normal.add_suffix('_std')
wine_normal.to_excel(r'E:\PCA\wine_normal.xlsx')

# Amount of variance explained by each PC
explained_var = pca.explained_variance_ratio_ # explained variance amount
explained_var_percent = np.round(explained_var,decimals=4)*100 # explained variance percentage
cum_var = np.cumsum(explained_var) # cumulative variance
cum_var_percent = np.cumsum(np.round(explained_var,decimals = 4)*100) # cumulative variance percentage

variance_explained_df = pd.DataFrame([explained_var,explained_var_percent,cum_var,cum_var_percent])
a = pd.Series(['explained_var','explained_var_percent','cum_var','cum_var_percent'])
variance_explained_df = pd.concat([a,variance_explained_df],axis=1)
variance_explained_df.columns = ['Variance','PC1','PC2','PC3','PC4','PC5','PC6','PC7','PC8','PC9','PC10','PC11','PC12','PC13']
variance_explained_df # details of explained and cumulative variance

# Variance plot for PCA components obtained 
plt.plot(cum_var,color="red")

# plot between Prinicipal Components
x = pca_scores[:,0]
pca_scores[:,1]
# x = np.array(pca_values[:,0])
y = pca_scores[:,1]
z = pca_scores[:,2]
plt.plot(x,y,'bo')
plt.plot(x,z,'bo')

''' Now we have obtained Principal components. First 9 PCs have explained 94% 
of variance of the original variables . So let us choose these 9 PCs. Let us run 
clustering machine learning algorithm first using all 13 variables and then using
these 9 PCs. Let us see if the reduced dimensions will give same result as 
original variables. '''

#################### Hierarchial Clustering using 13 Variables
wine_normal.head(3)
import scipy.cluster.hierarchy as sch

clust_dend = sch.linkage(wine_normal, method='complete', metric = 'euclidean')
%matplotlib qt
sch.dendrogram(clust_dend); # 3 or 4 clusters are clearly visible

from sklearn.cluster import AgglomerativeClustering
hclust_complete = AgglomerativeClustering(n_clusters=3,linkage='complete',affinity='euclidean').fit(wine_normal)
cluster_labels = pd.Series(hclust_complete.labels_)
wine_data.insert(0,'clust_complete',cluster_labels)
aggr_complete= wine_data.iloc[:,2:].groupby(wine_data.clust_complete).mean()

######################## Hierarchial clustering using 9 PC
pca_scores.columns
pca_scores_new = pca_scores.iloc[:,0:9] # selecting first 9 PCs (94% variance explained)
pca_scores_new.columns


clust_dend_pca = sch.linkage(pca_scores_new, method='complete', metric = 'euclidean')
%matplotlib qt
sch.dendrogram(clust_dend_pca); # 3 or 4 clusters are clearly visible

'''
PC = 9, then 3 or 4 clusters
PC =10, then 2 or 5 clusters
'''

hclust_complete_pc9 = AgglomerativeClustering(n_clusters=3,linkage='complete',affinity='euclidean').fit(pca_scores_new)
cluster_labels = pd.Series(hclust_complete_pc9.labels_)
wine_data.insert(0,'clust_complete_PC9',cluster_labels)
aggr_complete_PC9= wine_data.iloc[:,3:].groupby(wine_data.clust_complete_PC9).mean()
wine_data.columns

#plotting clusters 
import seaborn as sns
pca_scores_new.insert(0,'clust_complete_PC9',pd.Series(hclust_complete_pc9.labels_))
sns.lmplot(data=pca_scores_new, x='PC1', y='PC2',hue='clust_complete_PC9',
           fit_reg=False, legend=True, legend_out=True)
# only 2 clusters clearly separated, 3rd cluster samples(dots) are overlapping


pd.crosstab(wine_data.clust_complete, wine_data.clust_complete_PC9)
'''
clust_complete_PC9   0   1   2
clust_complete                
0                   60   9   0
1                   46  11   1
2                    2   3  46  '''

pd.crosstab(wine_data.clust_complete, wine_data.Type)
'''
Type             1   2   3
clust_complete            
0               51  18   0
1                8  50   0
2                0   3  48 '''

pd.crosstab(wine_data.clust_complete_PC9, wine_data.Type)
'''
Type                 1   2   3
clust_complete_PC9            
0                   59  48   1
1                    0  20   3
2                    0   3  44 '''
''' the clustering is not very good when use hierarchial clustering.
when compare aggregate means for 13 variables and 9 PCs, the means are 
different '''

#################### Kmeans Clustering using 13 Variables

from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=3, random_state=0)
kmeans.fit(wine_normal)
cluster_labels = pd.Series(kmeans.labels_)
wine_data.insert(0,'clust_kmeans',cluster_labels)
aggr_kmeans= wine_data.iloc[:,4:].groupby(wine_data.clust_kmeans).mean()


#################### Kmeans Clustering using first 9 PCs

kmeans_pc9 = KMeans(n_clusters=3, random_state=0)
kmeans_pc9.fit(pca_scores_new) # using 9 PCs
cluster_labels = pd.Series(kmeans_pc9.labels_)
wine_data.insert(0,'clust_kmeans_pc9',cluster_labels)
aggr_kmeans_pc9= wine_data.iloc[:,5:].groupby(wine_data.clust_kmeans_pc9).mean()

#plotting clusters 
pca_scores_new.insert(0,'clust_kmeans_pc9',cluster_labels)
%matplotlib qt
clus1 = sns.lmplot(data=pca_scores_new, x='PC1', y='PC2', hue='clust_kmeans_pc9', 
                   fit_reg=False, legend=True, legend_out=True)
# 3 clear clusters are visible


# Creating a Pandas Excel writer using XlsxWriter as the engine.
writer = pd.ExcelWriter('E:\PCA\pca.xlsx', engine='xlsxwriter')

# Wriing each aggregate output to a different worksheet.
aggr_complete.to_excel(writer, sheet_name='Sheet1')
aggr_complete_PC9.to_excel(writer, sheet_name='Sheet2')
aggr_kmeans.to_excel(writer, sheet_name='Sheet3')
aggr_kmeans_pc9.to_excel(writer,sheet_name='Sheet4')

writer.save() # closing writer and saving the file

import os
os.getcwd()

pd.crosstab(wine_data.clust_kmeans, wine_data.Type)
'''
Type           1   2   3
clust_kmeans            
0             59   3   0
1              0  65   0
2              0   3  48'''
pd.crosstab(wine_data.clust_kmeans_pc9, wine_data.Type)
# same result as above
pd.crosstab(wine_data.clust_kmeans,wine_data.clust_kmeans_pc9)
'''
clust_kmeans_pc9   0   1   2
clust_kmeans                
0                 62   0   0
1                  0  65   0
2                  0   0  51 '''
''' the aggregate means are same when we compared kmeans clsuters
using all 13 variables and 9 PCs '''


''' 
CONCLUSIONS

The wine dataset has 13 variables. For further analysis, using more
variables may reduce the performance. Hence we have performed principal
component analysis. The variance explained by 3 PCs was just 66% and 
5 PCs was just 80%. So we have used 9 Pcs which explains 94% variance.

Next we have performed clustering analysis to check the performance of
9 PCs. The PCs obtained have given good results with kmeans than with 
hierarchial clustering. 

A few principal components have been useful in explaining the variance
 in the dataset.
'''

File: ASSIGNMENT_PCA_Wine.R
##### ASSIGNMENT PCA using wine dataset

# Business Problem: Performing principal component analysis for dimension reduction

wine_data = read.csv("E:\\PCA\\wine.csv")
dim(wine_data)
str(wine_data)
table(wine_data$Type)
# we have 178 records and 14 variables. First variable 'Type' has 3 classes. Other
# variables are of numeric type. we will use these 13 numeric vars for PCA

wine <- wine_data[,-1]
head(wine,3)
attach(wine)

summary(wine_data)
# The variables have different ranges, so this will affect our analysis. We will use
# correlation rather than covariance matrix to get PCs.

cor(wine) # there exists correlation among the variables

pca <- princomp(wine,cor = TRUE, scores = TRUE, covmat = NULL)
str(pca)
''' 
gives details about standard deviation, center, number of records, call(function).
Also gives loadings (weights or correlation coefficients), scale and scores.
'''
summary(pca)
'''
  Comp.1    Comp.2    Comp.3    Comp.4     Comp.5     Comp.6    Comp.7     Comp.8     Comp.9 
0.3619885 0.5540634 0.6652997 0.7359900 0.80162293 0.85098116 0.89336795 0.92017544 0.94239698

# 3 PCs are ecplaining 67% of variance, 5 PCs are explaining 80% variance
# 9 PCs are explaining 94% variance, let us do further analysis using them '''

pca$sdev
pca$loadings
pca$scores
pca_scores <- pca$scores
loadings(pca)


sum(pca$loadings[,1]**2) # checking sum of squared weights is 1

plot(pca$scores[,1],pca$scores[,2])
cor(pca$scores[,1],pca$scores[,2])
cor(pca_scores) # now the cor of PCs is 0 . there existed correlation among variables

biplot(pca)
'''
# Bottom axis gives PC1 scores, left axis gives PC2 scores. 
#Top axis gives PC1 loadings and right axis gives PC2 loadings

# the variables close to X-axis influence PC1 the most. from the biplot we see
# that Phenols, proanthicyanins, flavonoids, non-flavonoids and alcalinity
# greatly influence PC1. 
# The variables close to Y-axis influence PC2. They are mainly color and ash

The angle between Phenols and Proanthocyanins is very less, indicating they have
positive correlation. whereas angle btw ash and Phenols is a right angle, so they
are not likely to be correlated. Angle between Hue and Malic is 180 degrees, 
they are negatively correlated.

'''
cor(Phenols,Proanthocyanins) #  0.6124131
cor(Ash,Phenols) # 0.1289795
cor(wine_data$Malic,wine_data$Hue) # -0.5612957

plot(cumsum(pca$sdev*pca$sdev)*100/(sum(pca$sdev*pca$sdev)),type="b")
plot(pca) # first 3 components are explaining much of the variance
plot(pca, type='l') # elbow plot also shows 3 clusters 

############ Hierarchial clustering with 13 variables
wine_normal <- scale(wine)
dist <- dist(wine_normal,method = 'euclidean')
clust_complete <- hclust(dist,method = 'complete')

# Dendrogram: to see how many clusters can be formed
plot(clust_complete,hang = -1,main='Dendrogram with all 13 variables') # showing 2 to 3 clusters
rect.hclust(clust_complete,k=3,border = 'blue')

# forming 3 clusters
groups <- cutree(clust_complete,3)
clust_hier_13vars <- as.matrix(groups)
finaldata <- cbind(clust_hier_13vars,wine_data)
View(aggregate(finaldata[,-c(1,2)],by=list(clust_hier_13vars),FUN=mean)) 

library(RColorBrewer)
library(scales)
palette(alpha(brewer.pal(9,'Set1'), 0.5))
plot(finaldata[,c(3,4)],col=clust_hier_13vars,pch=16,main='Hierarchial clustering with all 13 variables')
# when we visualize the clusters using combination of 2 variables, find overlapping clusters


############ Hierarchial clustering with 3 PCs
wine_3pc <- pca$scores[,1:3]
dist <- dist(wine_3pc,method = 'euclidean')
clust_complete1 <- hclust(dist,method = 'complete')

plot(clust_complete1,hang = -1,main="Dendrogram with 3 PCs") # showing 2 to 3 clusters
rect.hclust(clust_complete1,k=3,border = 'blue') # 3 irregular sized clusters formed

groups <- cutree(clust_complete1,3)
clust_hier_3pc <- as.matrix(groups)
finaldata <- cbind(clust_hier_3pc,finaldata)
View(aggregate(finaldata[,-c(1,2,3)],by=list(clust_hier_3pc),FUN=mean)) 
plot(wine_3pc,col=clust_hier_3pc,pch=16,main='Hierarchial clustering with 3 PCs')
# 2 clusters are dinstinct, but one cluster is overlapping

############ Hierarchial clustering with 5 PCs
wine_5pc <- pca$scores[,1:5]
dist <- dist(wine_5pc,method = 'euclidean')
clust_complete2 <- hclust(dist,method = 'complete')

plot(clust_complete2,hang = -1) # showing 2 to 5 clusters
rect.hclust(clust_complete2,k=3,border = 'blue') # 3 irregular sized clusters formed

groups <- cutree(clust_complete2,3)
clust_hier_5pc <- as.matrix(groups)
finaldata <- cbind(clust_hier_5pc,finaldata)
View(aggregate(finaldata[,-c(1,2,3,4)],by=list(clust_hier_5pc),FUN=mean)) 

plot(wine_5pc,col=clust_hier_5pc,pch=16,main='Hierarchial clustering with 5 PCs')
# 2 clusters are dinstinct, but one cluster is overlapping, similar to 3-PC clustering

############ Hierarchial clustering with 9 PCs
wine_9pc <- pca$scores[,1:9]
dist <- dist(wine_9pc,method = 'euclidean')
clust_complete3 <- hclust(dist,method = 'complete')

plot(clust_complete3,hang = -1) # showing 2 to 5 clusters
rect.hclust(clust_complete3,k=3,border = 'blue') # 3 irregular sized clusters formed

groups <- cutree(clust_complete3,3)
clust_hier_9pc <- as.matrix(groups)
finaldata <- cbind(clust_hier_9pc,finaldata)
View(aggregate(finaldata[,-c(1,2,3,4,5)],by=list(clust_hier_9pc),FUN=mean)) 

plot(wine_9pc,col=clust_hier_9pc,pch=16,main='Hierarchial clustering with 9 PCs')
# 2 clusters are dinstinct, but one cluster is overlapping, similar to 3-PC 
# and 5-PC clustering


## The clusters of hierarchial clustering are not properly separable (neither in
# original data nor using a few PCs).

############### Kmeans clustering using all variables
library(animation)
set.seed(123)
kmeans_13vars <- kmeans(wine_normal,3)
kmeans.ani(wine_normal,3) # shows for 2 variables, but clusters are overlapping
clust_kmeans_13vars <- kmeans_13vars$cluster

finaldata <- cbind(clust_kmeans_13vars,finaldata)
View(aggregate(finaldata[,-c(1:6)],by=list(clust_kmeans_13vars),FUN = mean))

plot(finaldata[,c(7,8)],col=clust_kmeans_13vars,pch=16,main='Kmeans with all 13 variables')
# using 2 variables if plot the clusters, find 3 overlapping clusters  

############### Kmeans clustering using 3 PCs
head(wine_3pc)
kmeans_3pc <- kmeans(wine_3pc,3)
kmeans.ani(wine_3pc,3)
clust_kmeans_3pc <- kmeans_3pc$cluster
finaldata <- cbind(clust_kmeans_3pc,finaldata)
View(aggregate(finaldata[,-c(1:7)],by=list(clust_kmeans_3pc),FUN = mean))

plot(wine_3pc,col=clust_kmeans_3pc,pch=16,main='Kmeans with 3 PCs')
# 3 clear clusters are visible

############### Kmeans clustering using 5 PCs
head(wine_5pc)
kmeans_5pc <- kmeans(wine_5pc,3)
kmeans.ani(wine_5pc,3)
clust_kmeans_5pc <- kmeans_5pc$cluster
finaldata <- cbind(clust_kmeans_5pc,finaldata)
View(aggregate(finaldata[,-c(1:8)],by=list(clust_kmeans_5pc),FUN = mean))

plot(wine_5pc,col=clust_kmeans_5pc,pch=16,main='Kmeans with 5 PCs')
# 3 clear clusters are visible

############### Kmeans clustering using 9 PCs
head(wine_9pc)
kmeans_9pc <- kmeans(wine_9pc,3)
kmeans.ani(wine_9pc,3)
clust_kmeans_9pc <- kmeans_9pc$cluster
finaldata <- cbind(clust_kmeans_9pc,finaldata)
View(aggregate(finaldata[,-c(1:9)],by=list(clust_kmeans_9pc),FUN = mean))

plot(wine_9pc,col=clust_kmeans_9pc,pch=16,main='Kmeans with 9 PCs')
# 3 clear clusters are visible

'''
Though the variance explained by 3 PCs is just 67%, 5-PCs is 80% and
9-PCs is 94%, the clusters formed by using these PCs is similar. '''

'''
CONCLUSIONS

The wine dataset has many variables. We have used principal component analysis
to reduce the dimensions. To know if the principal components will perform better,
we have used clustering on original dataset and then on PC scores.

We have done hierarchial and kmeans clustering using 3,5 and 9 PCs and all
13 variables. The PCs have produced results similar to that of original 
variables.

Principal components have been successful in capturing the essence of the data.
'''




